api/controllers/UserController.js

```diff
@@ -234,6 +235,318 @@ module.exports = {
     });
   },
 
+  permaBan: function (req, res) {
+    /*  Form parameters:
+          req.params.username: The user who is being banned
+          req.params.banNote: The ban reason to go on the mod log (not visible to banned user, 300 characters max)
+          req.params.banMessage: The note that gets sent with the "you are banned" PM
+          req.params.banlistEntry: The ban reason to appear on the public banlist
+          req.params.duration: The number of days that the user will be banned for.
+        Permaban process:
+          1. Ban user from /r/pokemontrades
+          2. Ban user from /r/SVExchange
+          3. Add "BANNED USER" to user's flair on /r/pokemontrades
+          4. Add user's friend code to /r/pokemontrades AutoModerator config (2 separate lists)
+          5. Add user's friend code to /r/SVExchange AutoModerator config (2 separate lists)
+          6. Remove all of the user's TSV threads on /r/SVExchange
+          7. Add user's info to banlist wiki on /r/pokemontrades
+    */
+    if (!req.user.isMod) {
+      res.json("Not a mod", 403);
+      return;
+    }
+    req.params = req.allParams();
+    if (!req.params.username) {
+      res.json("No username", 400);
+      return;
+    }
+    if (req.params.banNote.length > 300) {
+      res.json("Ban note too long", 400);
+    }
+    try {
+      var duration = req.params.duration ? parseInt(req.params.duration) : 0;
+      if (duration < 0) {
+        res.json("Invalid duration", 400);
+      }
+    } catch (err) {
+      res.json("Invalid duration", 400);
+    }
+
+    var number_of_tasks = duration ? 2 : 8;
+    var completed_tasks = 0;
+
+    //Ban user from the two subs
+    var banFromSub = function (subreddit) {
```

Comment:
Please, please, please, extract these out soemwhere. This file is far too big now, and having all of these functions in it makes me sad.